lib/src/modules/task.cc

```diff
@@ -355,7 +363,10 @@ static fs::path getCachedTaskFile(const fs::path& task_cache_dir,
     LOG_DEBUG("Verifying task file based on {1}", file.toString());
 
     try {
-        auto cache_dir = createCacheDir(task_cache_dir, file.get<std::string>("sha256"));
+        auto cache_dir = [&]() {
+            pcp_util::lock_guard<pcp_util::mutex> the_lock { task_cache_dir_mutex };
+            return createCacheDir(task_cache_dir, file.get<std::string>("sha256"));
+        }();
         return updateTaskFile(master_uris, client, cache_dir, file);
```

Comment:
Does this line need to be part of the critical section above? It seems possible for the purge and this code to run concurrently. In that case, the following sequence of operations could happen:

A) The purge function notices the task file needs to be deleted
B) The running task operation checks that the task file exists and sees it does
C) The purge function deletes the task file
D) The running task operation presumably explodes

Reply:
Purge should only operate on directories. A side-effect of createCacheDir is to update mtime of the directory. We then download files into that directory. The purge should only: check whether the directory is old, and if so remove the directory. So I believe it behaves correctly.